File: learnSphere/frontend/learnsphere/src/app/features/student/services/messaging.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from '@core/services/api.service';
import { ConversationDto, MessageDto, StartConversationRequest } from '@core/models/social.model';

/**
 * MessagingService (student) — real HTTP calls against the shared
 * /api/conversations backend (usable by any authenticated role).
 */
@Injectable({
  providedIn: 'root'
})
export class MessagingService {

  constructor(private api: ApiService) {}

  // Returns all conversations for the current student.
  getConversations(): Observable<ConversationDto[]> {
    return this.api.get<ConversationDto[]>('/conversations');
  }

  // Returns the message thread for a conversation (marks messages read as a side effect).
  getMessages(conversationId: number): Observable<MessageDto[]> {
    return this.api.get<MessageDto[]>(`/conversations/${conversationId}/messages`);
  }

  // Sends a message in an existing conversation.
  sendMessage(conversationId: number, text: string): Observable<MessageDto> {
    return this.api.post<MessageDto>(`/conversations/${conversationId}/messages`, { text });
  }

  // Starts (or reuses) a conversation with another user, optionally scoped to a course.
  startConversation(request: StartConversationRequest): Observable<ConversationDto> {
    return this.api.post<ConversationDto>('/conversations', request);
  }
}

File: learnsphere-backend/src/main/java/com/MOOC/OnlineLearningPlatfrom/Service/MessagingService.java
package com.MOOC.OnlineLearningPlatfrom.Service;

import com.MOOC.OnlineLearningPlatfrom.Dto.ConversationResponseDto;
import com.MOOC.OnlineLearningPlatfrom.Dto.MessageResponseDto;
import com.MOOC.OnlineLearningPlatfrom.Dto.SendMessageRequestDto;
import com.MOOC.OnlineLearningPlatfrom.Dto.StartConversationRequestDto;
import com.MOOC.OnlineLearningPlatfrom.Security.CustomUserDetails;

import java.util.List;

public interface MessagingService {
    List<ConversationResponseDto> getConversations(CustomUserDetails principal);
    List<MessageResponseDto> getMessages(Long conversationId, CustomUserDetails principal);
    MessageResponseDto sendMessage(Long conversationId, SendMessageRequestDto request, CustomUserDetails principal);
    ConversationResponseDto startConversation(StartConversationRequestDto request, CustomUserDetails principal);
}
